File: AkwaierMAC_IOS/AkwaierKern/Sources/AkwaierKern/Scorelijst.swift
import Foundation

/// Eén regel uit de scorelijst.
public struct Scoreregel: Codable, Equatable {
    public var naam: String
    public var vermogen: Int

    public init(naam: String, vermogen: Int) {
        self.naam = naam
        self.vermogen = vermogen
    }
}

/// De inhoud van het scorebestand: de gekozen taal en de ranglijst.
public struct Scorebestand: Codable {
    /// "nl" of "en"; leeg als er nog nooit een taal gekozen is.
    public var taal: String = ""
    public var scores: [Scoreregel] = []

    public init() {}
}

/// Vervangt het bestand "SCORE" uit het origineel (regels 4050-4195) door een
/// leesbaar JSON-bestand. Behalve de ranglijst staat ook de taalkeuze erin,
/// zodat het spel de volgende keer in dezelfde taal begint.
///
/// Op Windows staat dit bestand naast het programma. Dat kan op iOS niet, dus
/// hier gaat het naar Application Support — op de iPad de enige plek waar een
/// app mag schrijven, en op de Mac de gebruikelijke.
public enum Scorelijst {

    private static let bewaar = 30

    public static var pad: URL = {
        let map = (try? FileManager.default.url(for: .applicationSupportDirectory,
                                                in: .userDomainMask,
                                                appropriateFor: nil, create: true))
            ?? URL(fileURLWithPath: NSTemporaryDirectory())
        let eigen = map.appendingPathComponent("Akwaier", isDirectory: true)
        try? FileManager.default.createDirectory(at: eigen, withIntermediateDirectories: true)
        return eigen.appendingPathComponent("akwaier-scores.json")
    }()

    /// Leest het bestand. Een bestand van vóór de taalkeuze bevat alleen een
    /// lijst; dat wordt nog gewoon gelezen.
    public static func lees() -> Scorebestand {
        guard let data = try? Data(contentsOf: pad) else { return Scorebestand() }
        if let bestand = try? JSONDecoder().decode(Scorebestand.self, from: data) {
            return bestand
        }
        if let lijst = try? JSONDecoder().decode([Scoreregel].self, from: data) {
            var bestand = Scorebestand()
            bestand.scores = lijst
            return bestand
        }
        // Een stukgelopen of met de hand aangepast bestand mag het spel niet
        // tegenhouden; de lijst begint dan gewoon opnieuw.
        return Scorebestand()
    }

    private static func schrijf(_ bestand: Scorebestand) {
        let codeerder = JSONEncoder()
        codeerder.outputFormatting = [.prettyPrinted, .sortedKeys]
        guard let data = try? codeerder.encode(bestand) else { return }
        try? data.write(to: pad, options: .atomic)
    }

    @discardableResult
    public static func bijwerken(_ nieuw: [(naam: String, vermogen: Int)]) -> [Scoreregel] {
        var bestand = lees()
        bestand.scores.append(contentsOf: nieuw.map {
            Scoreregel(naam: $0.naam, vermogen: $0.vermogen)
        })
        bestand.scores.sort { $0.vermogen > $1.vermogen }
        if bestand.scores.count > bewaar { bestand.scores = Array(bestand.scores.prefix(bewaar)) }
        bestand.taal = Taal.code
        schrijf(bestand)
        return bestand.scores
    }

    /// Zet de taal op de bewaarde keuze. Staat er nog niets in het bestand, dan
    /// beslist de taalinstelling van het apparaat.
    public static func laadTaal() {
        let code = lees().taal
        if code.trimmingCharacters(in: .whitespaces).isEmpty {
            Taal.engels = Taal.apparaatIsEngels()
        } else {
            Taal.zetCode(code)
        }
    }

    public static func bewaarTaal() {
        var bestand = lees()
        guard bestand.taal != Taal.code || !FileManager.default.fileExists(atPath: pad.path) else { return }
        bestand.taal = Taal.code
        schrijf(bestand)
    }
}

File: Akwaier/AkwaierWin/Engine/Scorelijst.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Akwaier.Engine;

/// <summary>Eén regel uit de scorelijst.</summary>
public sealed class Scoreregel
{
    [JsonPropertyName("naam")] public string Naam { get; set; } = "";
    [JsonPropertyName("vermogen")] public int Vermogen { get; set; }
}

/// <summary>De inhoud van het scorebestand: de gekozen taal en de ranglijst.</summary>
public sealed class Scorebestand
{
    /// <summary>"nl" of "en"; leeg als er nog nooit een taal gekozen is.</summary>
    [JsonPropertyName("taal")] public string TaalCode { get; set; } = "";
    [JsonPropertyName("scores")] public List<Scoreregel> Scores { get; set; } = new();
}

/// <summary>
/// Vervangt het bestand "SCORE" uit het origineel (regels 4050-4195) door een
/// leesbaar JSON-bestand naast het programma. Behalve de ranglijst staat ook de
/// taalkeuze erin, zodat het spel de volgende keer in dezelfde taal begint.
/// </summary>
public static class Scorelijst
{
    private const int Bewaar = 30;

    private static readonly JsonSerializerOptions Opties = new() { WriteIndented = true };

    public static string Pad => Path.Combine(
        Path.GetDirectoryName(Environment.ProcessPath) ?? ".", "akwaier-scores.json");

    /// <summary>
    /// Leest het bestand. Een bestand van vóór de taalkeuze bevat alleen een
    /// lijst; dat wordt nog gewoon gelezen.
    /// </summary>
    public static Scorebestand Lees()
    {
        try
        {
            if (!File.Exists(Pad)) return new Scorebestand();
            string tekst = File.ReadAllText(Pad);
            using var doc = JsonDocument.Parse(tekst);
            if (doc.RootElement.ValueKind == JsonValueKind.Array)
                return new Scorebestand
                {
                    Scores = JsonSerializer.Deserialize<List<Scoreregel>>(tekst) ?? new(),
                };
            return JsonSerializer.Deserialize<Scorebestand>(tekst) ?? new Scorebestand();
        }
        catch (Exception)
        {
            // Een stukgelopen of met de hand aangepast bestand mag het spel niet
            // tegenhouden; de lijst begint dan gewoon opnieuw.
            return new Scorebestand();
        }
    }

    private static void Schrijf(Scorebestand bestand)
    {
        try { File.WriteAllText(Pad, JsonSerializer.Serialize(bestand, Opties)); }
        catch (Exception) { /* alleen-lezen map: dan geen lijst, maar wel doorspelen */ }
    }

    public static List<Scoreregel> Bijwerken(IEnumerable<(string Naam, int Vermogen)> nieuw)
    {
        var bestand = Lees();
        bestand.Scores.AddRange(nieuw.Select(n => new Scoreregel
        {
            Naam = n.Naam,
            Vermogen = n.Vermogen,
        }));
        bestand.Scores = bestand.Scores.OrderByDescending(r => r.Vermogen).Take(Bewaar).ToList();
        bestand.TaalCode = Taal.Code;
        Schrijf(bestand);
        return bestand.Scores;
    }

    /// <summary>
    /// Zet de taal op de bewaarde keuze. Staat er nog niets in het bestand, dan
    /// beslist de taalinstelling van Windows.
    /// </summary>
    public static void LaadTaal()
    {
        string code = File.Exists(Pad) ? Lees().TaalCode : "";
        if (string.IsNullOrWhiteSpace(code)) Taal.Engels = Taal.PcIsEngels();
        else Taal.ZetCode(code);
    }

    public static void BewaarTaal()
    {
        var bestand = Lees();
        if (bestand.TaalCode == Taal.Code && File.Exists(Pad)) return;
        bestand.TaalCode = Taal.Code;
        Schrijf(bestand);
    }
}
